lib/spack/spack/spack_deps.py

```diff
@@ -0,0 +1,153 @@
+            {'prefix': python_prefix, 'spec': str(external_python)}
+        ]
+    }
+
+    with spack.config.override('packages:python::', entry):
+        yield
+
+
+def make_module_available(module, spec=None, install=False):
+    """Ensure module is importable"""
+    # If we already can import it, that's great
+    try:
+        __import__(module)
+        return
+    except ImportError:
+        pass
+
+    # If it's already installed, use it
+    # Search by spec
+    spec = spack.spec.Spec(spec or module)
+
+    # We have to run as part of this python
+    # We can constrain by a shortened version in place of a version range
+    # because this spec is only used for querying or as a placeholder to be
+    # replaced by an external that already has a concrete version. This syntax
+    # is not suffucient when concretizing without an external, as it will
+    # concretize to python@X.Y instead of python@X.Y.Z
+    spec.constrain('^python@%d.%d' % sys.version_info[:2])
+    installed_specs = spack.store.db.query(spec, installed=True)
+
+    for ispec in installed_specs:
+        # TODO: make sure run-environment is appropriate
+        module_path = os.path.join(ispec.prefix,
+                                   ispec['python'].package.site_packages_dir)
+        module_path_64 = module_path.replace('/lib/', '/lib64/')
+        try:
+            sys.path.append(module_path)
+            sys.path.append(module_path_64)
+            __import__(module)
+            return
+        except ImportError:
+            tty.warn("Spec %s did not provide module %s" % (ispec, module))
+            sys.path = sys.path[:-2]
+
+    if not install:
+        raise Exception  # TODO specify
+
+    with system_python_context():
+        # We will install for ourselves, using this python if needed
+        # Concretize the spec
+        spec.concretize()
+    spec.package.do_install()
+
+    module_path = os.path.join(spec.prefix,
+                               spec['python'].package.site_packages_dir)
+    module_path_64 = module_path.replace('/lib/', '/lib64/')
+    try:
+        sys.path.append(module_path)
+        sys.path.append(module_path_64)
+        __import__(module)
+        return
+    except ImportError:
+        sys.path = sys.path[:-2]
+        raise Exception  # TODO: specify
+
+
+def get_executable(exe, spec=None, install=False):
+    """Find an executable named exe, either in PATH or in Spack
+
+    Args:
+        exe (str): needed executable name
+        spec (Spec or str): spec to search for exe in (default exe)
+        install (bool): install spec if not available
+
+    When ``install`` is True, Spack will use the python used to run Spack as an
+    external. The ``install`` option should only be used with packages that
+    install quickly (when using external python) or are guaranteed by Spack
+    organization to be in a binary mirror (clingo)."""
+    # Easy, we found it externally
+    # TODO: Add to externals/database?
```

Comment:
For this TODO, I am wondering if we should perform installations of Spack dependencies in a separate store - in particular if we proceed reusing as much as possible installed packages with the ASP based concretizer. A use case I have in mind, for instance, is somebody who wants to uninstall everything:
```console
$ spack uninstall -ay
```
This in my opinion shouldn't uninstall packages that are needed by Spack to run, and shouldn't make a re-build of those dependencies necessary later. In #20068 I used environments to obtain some segregation, but I understand that we may not want to use that mechanism.

Along the same lines, I am not sure we should show `clingo` or `flake8` etc. when a:
```console
$ spack find
```
command is given, or at least not the ones used for bootstrapping Spack.

Reply:
I think the user should be able to uninstall things that Spack build to bootstrap itself, but we should add a dependent on those packages (from Spack) so that they cannot be uninstalled without the `-f` option. But (for example) Spack will bootstrap different specs depending on the python Spack runs with, and the user may want to remove some of those packages before changing pythons in their default environment (and bootstrapping new versions).